### PR TITLE
Fix string representation of Gi binary format when specifying in memory limits or requests. 

### PR DESCRIPTION
**Pull request by aveshagarwal**
**What this PR does / why we need it**:
This PR fixes an issue when memory limit or request is specified in pod spec in Gi format, it gets converted to milli byte format, which looks rather weird. For example:

```
          limits:
            memory: 1.8Gi
```

gets converted to:

```
      limits:
            memory: 1932735283200m
```
When the pod is described or  when the pod's json or yaml format is output. 

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

@derekwaynecarr

### CODE REVIEW

File: staging/src/k8s.io/apimachinery/pkg/api/resource/quantity_test.go [hunk 1]
```diff
@@ -736,15 +736,128 @@ func TestQuantityParseEmit(t *testing.T) {
...
+		{".000001M", "1"},
+		{".000000001M", "1m"},
+		{".000000000001M", "1u"},
+
+		// Mi (>1)
 		{"1Mi", "1Mi"},
-		{"1Gi", "1Gi"},
+		{"1.23Mi", "1.23Mi"},
 		{"1024Mi", "1Gi"},
-		{"1000M", "1G"},
```
Comment (smarterclayton) [inline staging/src/k8s.io/apimachinery/pkg/api/resource/quantity_test.go]: Where did this test go?

Comment (smarterclayton) [inline staging/src/k8s.io/apimachinery/pkg/api/resource/quantity_test.go]: If this doesn't work anymore, this code change is not safe.

Comment (aveshagarwal) [inline staging/src/k8s.io/apimachinery/pkg/api/resource/quantity_test.go]: I have added these numbers back, not sure how I missed them. If you have any other numbers in mind, let me know.

**Comment by aveshagarwal** [comment]
@smarterclayton  anything else you think is not addressed?

File: staging/src/k8s.io/apimachinery/pkg/api/resource/quantity_test.go [hunk 2]
```diff
@@ -736,15 +736,134 @@ func TestQuantityParseEmit(t *testing.T) {
...
+		{"1836Ki", "1836Ki"},
+
+		// Ki (<1)
+		{".2Ki", ".2Ki"},
+		{".5Ki", ".5Ki"},
+		{".02Ki", ".02Ki"},
+		{".023Ki", ".023Ki"},
+		{".0234Ki", ".0234Ki"},
+		{".0004Ki", ".0004Ki"},
+		{".001Ki", ".001Ki"},
```
Comment (smarterclayton) [inline staging/src/k8s.io/apimachinery/pkg/api/resource/quantity_test.go]: This was the actual bug (that is fixed here).  Before `.001Ki` was being normalized to `1024m` which is not correct.  `.001 * 1024 != 1024m`

File: staging/src/k8s.io/apimachinery/pkg/api/resource/quantity_test.go [hunk 3]
```diff
@@ -736,15 +736,134 @@ func TestQuantityParseEmit(t *testing.T) {
...
+		{".2Mi", ".2Mi"},
+		{".5Mi", ".5Mi"},
+		{".02Mi", ".02Mi"},
+		{".023Mi", ".023Mi"},
+		{".0234Mi", ".0234Mi"},
+		{".0004Mi", ".0004Mi"},
+		{".001Mi", ".001Mi"},
+		{".000001Mi", ".000001Mi"},
+		{".000000001Mi", ".000000001Mi"},
+		{".000000000001Mi", "1049n"},
```
Comment (smarterclayton) [inline staging/src/k8s.io/apimachinery/pkg/api/resource/quantity_test.go]: Is this actually correct?

Comment (aveshagarwal) [inline staging/src/k8s.io/apimachinery/pkg/api/resource/quantity_test.go]: .000000000001Mi = .000000000001 x 1024 x 1024 = 10^-9 x 1048576 ~ 10^-9 x 1049 x 10^3  = 1049 x 10^-6 = 1049 x 10^-6 x 10^n=1049n

File: staging/src/k8s.io/apimachinery/pkg/api/resource/quantity_test.go (hunk 2 above)
Comment (aveshagarwal) [inline staging/src/k8s.io/apimachinery/pkg/api/resource/quantity_test.go]: .001Ki = .001 x 1024 = .001x 1024  x 1000 m = 1024m

Comment (derekwaynecarr) [inline staging/src/k8s.io/apimachinery/pkg/api/resource/quantity_test.go]: @aveshagarwal i dont follow this comment, but I agree that what was there previously was not correct.

Comment (aveshagarwal) [inline staging/src/k8s.io/apimachinery/pkg/api/resource/quantity_test.go]: I meant both 1024m and 0.001Ki are correct just different representations in different units.

Comment (smarterclayton) [reply staging/src/k8s.io/apimachinery/pkg/api/resource/quantity_test.go]: Hrm, if they are both correct, the canonical representation would be 1024m.  Normalization is supposed to remove decimals.  We show the highest non-decimal representation.

Comment (aveshagarwal) [reply staging/src/k8s.io/apimachinery/pkg/api/resource/quantity_test.go]: Ki format is only used for memory/storage in kubernetes.  Do we really want to show memory/storage in milli (1024m here) format?

**Comment by smarterclayton** [comment]
This was a spec issue, so it may not matter.  See if this is described in
the resource quantity proposal (would be under canonicalization).


On Sep 2, 2017, at 11:57 AM, Avesh Agarwal <notifications@github.com> wrote:

*@aveshagarwal* commented on this pull request.
------------------------------

In staging/src/k8s.io/apimachinery/pkg/api/resource/quantity_test.go
<https://github.com/kubernetes/kubernetes/pull/49732#discussion_r136698448>:

>  		{"1Ki", "1Ki"},
+		{"1.23Ki", "1.23Ki"},
+		{"1000Ki", "1000Ki"},
+		{"1024Ki", "1Mi"},
+		{"1224Ki", "1224Ki"},
+		{"1536Ki", "1536Ki"},
+		{"1836Ki", "1836Ki"},
+
+		// Ki (<1)
+		{".2Ki", ".2Ki"},
+		{".5Ki", ".5Ki"},
+		{".02Ki", ".02Ki"},
+		{".023Ki", ".023Ki"},
+		{".0234Ki", ".0234Ki"},
+		{".0004Ki", ".0004Ki"},
+		{".001Ki", ".001Ki"},

Ki format is only used for memory/storage in kubernetes. Do we really want
to show memory/storage in milli (1024m here) format?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/49732#discussion_r136698448>,
or mute the thread

**Comment by aveshagarwal** [comment]
>This was a spec issue, so it may not matter.  See if this is described in
the resource quantity proposal (would be under canonicalization).

I am not disagreeing regarding canonicalization. My point is as far as the numbers that are relevant to Kubernetes use cases are displayed correctly, we should be fine. There are several very small numbers in unit test cases (as they were already existed in unit test cases, so I took them too), are not really practical from Kubernetes use cases. This PR fixes large numbers (and other numbers) relevant to Kubernetes.

**Comment by smarterclayton** [comment]
Again, a fix would need to be correct here if we previously had
defined such a canonical behavior.

**Comment by aveshagarwal** [comment]
> Again, a fix would need to be correct here if we previously had
defined such a canonical behavior.

But if we go by that requirement, that means the current behavior is correct that means 1.8Gi is rightly converted to 1932735283200m because 1932735283200m is the closest integer we can get to for 1.8Gi.

**Comment by sjenning** [comment]
@smarterclayton @aveshagarwal @derekwaynecarr where did we end up on this?

/retest

File: staging/src/k8s.io/apimachinery/pkg/api/resource/quantity_test.go (hunk 2 above)
Comment (smarterclayton) [reply staging/src/k8s.io/apimachinery/pkg/api/resource/quantity_test.go]: That is the canonical representation.

**Comment by smarterclayton** [comment]
the behavior of canonicalization is that we return the smallest non-fractional number without losing precision:

https://github.com/kubernetes/kubernetes/pull/1185/files#diff-10d9c300e4d2054fa473090899a29f0aR59

0.001Ki is not correct - this PR would need to be changed to not incorrectly canonicalize that value.

**Comment by smarterclayton** [comment]
Although I don't know if we define negative mantissa units for Ki

File: staging/src/k8s.io/apimachinery/pkg/api/resource/quantity_test.go (hunk 2 above)
Comment (smarterclayton) [reply staging/src/k8s.io/apimachinery/pkg/api/resource/quantity_test.go]: Unless we declare it is not fractional, in which case the canonicalization is meaningless.

**Pull request closed without merging**